### ISSUE TITLE
bidi support dependencies/flag

### ISSUE THREAD

**Issue by cjolif** [ibm-js/deliteful#462]
After some investigation it happens we have some "bidi"-type features under bidi flag and some are not. We need to decide either to make 2 clear categories one under flag, one without flag or to come up with a single category which is managed in a consistent manner.

We should also decide what the proper naming for the flag is.

-----
Initial issue:

StarRating is doing those requires:

```js
	"requirejs-dplugins/has!bidi?./StarRating/bidi/StarRating",
	"requirejs-dplugins/has!bidi?delite/theme!./StarRating/themes/{{theme}}/StarRating_rtl.css"
```

It would probably better better to centralize all the bidi stuff in one module (here StarRating/bidi/StarRating) and thus require the CSS from this module instead of from the regular StarRating module.

**Comment by wkeese** [comment]
I agree the current code is wrong, but aren't widgets supposed to support RTL mode even when `has("bidi")` is false?   Otherwise, `delite/Widget#isLeftToRight()` would always just return "ltr".

If you want to change the meaning of `has("bidi")` to include any/all support of RTL text, then we should be consistent w/that across the code base.   But in that case it should be called `has("rtl")` not `has("bidi")`.

**Comment by cjolif** [comment]
I think it is already somewhat consistent in the sense that I think rtl CSS are already loaded only when `has("bidi")`  is true. Now maybe this was done by mistake instead of one purpose. This might well be the case as in some cases rtl styling is not in a separated CSS file. cc: @pruzand

**Comment by wkeese** [comment]
Yes, ProgressBar, SidePane, StarRating, and List have separate CSS files for RTL mode, which are only loaded when `has("bidi")` is true. 

OTOH, CheckBox, ComboBox, Switch, and delite itself support RTL mode regardless of the `has("bidi")` flag.

So, I guess you could call that "somewhat consistent", but let's make it truly consistent.


> If you want to change the meaning of `has("bidi")` to include any/all support of RTL text, then we should be consistent w/that across the code base.   But in that case it should be called `has("rtl")` not `has("bidi")`.

To be clear, "bidi" actually means "bidirectional text" which implies that the widget has code for dealing with single strings containing both Hebrew/Arabic and other scripts (like English, French, or Japanese).  It's a much more esoteric case than just displaying Hebrew or Arabic.

**Comment by wkeese** [comment]
To summarize from our meeting (as I remember it):

1. In both dojo and ibm-js, lots of code (ex: dojox/charting and StarRating) doesn't support RTL at all unless has("bidi") is enabled.   I.E. drawing the chart axis starting at the bottom right (instead of bottom left), or rendering StarRating's stars starting from the right to the left.   Other code like delite tries to support simple RTL even when has("bidi") is false, and uses has("bidi") only to enable advanced features like the textdir parameter
2. We could three options to make this consistent:
     * one has() flag turns on basic RTL support, and another has() flag turns on support for advanced features like the `textdir` parameter
     * one has() flag turns on everything
     * simple RTL support enabled all the time, but advanced features controlled by a has() flag
3. @cjolif will talk to @semion1956 about which is best.  If all of the middle eastern customers need full support (`textdir` property etc.) then maybe we should just have a single flag to control everything.  In that case we should think of a better name, because has("bidi") seems misleading.

**Comment by semion1956** [comment]
We have considered ```has("bidi")``` flag as a switch to support any advanced bidi features, not only *textdir*. Example is mirroring of charts, which from the one hand doesn't have native support and require additional code, and from the other hand, depends on specific locale, specific customer and even specific application: sometimes it is required, sometimes only desirable, and sometimes not needed at all. Advanced features aren't always required at the same time, however we don't think, that  user should have the separate flag for each of them.  *"Simple RTL support enabled all the time, but advanced features controlled by a has() flag"* is the best option from our point of view.
What about name - we agree with any decision. *"bidi"* we also like. Something like "advancedBidi" is more precisely, but longer.

**Comment by cjolif** [comment]
@semion1956 it seems to make sense, the question is now to categorize what we have in deliteful to be either "advanced" (under flag) feature vs not advanced.

If you grep for has("bidi") in deliteful do you consider some of these should not be under bidi flag? (for example the specific RTL CSS sometimes?)

**Comment by wkeese** [comment]
> the question is now to categorize what we have in deliteful to be either "advanced" (under flag) feature vs not advanced.

Indeed.  I was surprised to see @semion1956's comment:

> advanced bidi features ... Example is mirroring of charts

I thought mirroring of charts would be considered "simple".

So, looking forward to your definition of simple vs. advanced.  Hopefully you can define them from the application's perspective rather than about implementation details.

**Comment by semion1956** [comment]
Speaking of simple (or basic) bidi functionality, first of all we mean performing the globalization requirement - support of RTL GUI direction. We exclude from here the cases, when this, from the one hand, is not always required even in RTL environment, and from the other hand, requires not trivial implementation. From this point of view RTL-specific CSS in general should not be loaded under bidi flag (like this is done in List, ProgressBar, SidePane and StarRating). What about additional functionality like RTL-specific behavior of left/right arrow keys, it can be considered as advanced bidi support, but here - in my opinion - all depends on complexity of implementation.  I think, that in case of StarRating corresponding code can be safely included into the base class.

**Comment by wkeese** [comment]
> ... support of RTL GUI direction [without the `has("bidi")` flag]...

Thanks for the response. 

> We exclude from here the cases ... requires not trivial implementation. 

Unfortunately that makes things harder to document, and may lead to confusion.  (If 90% of widgets render correctly without `has("bidi")` , developers will probably expect 100% of them to render correctly.)   
Generally, complex RTL support code just indicates that the widget was coded badly (see for example #454).

But perhaps complex things like charting need non-trivial implementations.  So perhaps the exception is necessary from a practical standpoint.

**Comment by wkeese** [comment]
I'm working on this, specifically using separate files only for advanced bidi support, which in practice means support for `textdir`.   Changes in #521.

**Comment by wkeese** [comment]
In 86ec1850b6f15e53f459d10bf420ca50603ea7f7 I wrote "refs #462" but I guess it actually fixes #462 (i.e. this ticket).

IIRC we decided to stick with the name has("bidi") to indicate support for advanced bidi features, and to support RTL out of the box.   And I think those changes are finished.
